src/openms/include/OpenMS/METADATA/ID/IdentificationData.h

```diff
@@ -523,6 +505,35 @@ namespace OpenMS
     /// Get range of matches (cf. @p equal_range) for a given observation
     std::pair<ObservationMatchRef, ObservationMatchRef> getMatchesForObservation(ObservationRef obs_ref) const;
 
+    /*!
+      @brief Helper function for filtering observation matches (e.g. PSMs) in IdentificationData
+
+      Depending on parameter @p cleanup_affected, the data structure may be cleaned up (IdentificationData::cleanup) to remove any invalidated references at the end of this operation.
+
+      @param func Functor that returns true for container elements to be removed
+      @param cleanup_affected Will filtering invalidate other parts of @p id_data that need to be cleaned up?
```

Comment:
I'm not sure about the utility of this parameter. I think it would be better (in all of the functions below) to count elements before and after the `removeFromSetIf_(...)`, and always call `cleanup()` if any were removed.

Reply:
On second thought, in some cases it may be useful to have a `defer_cleanup` option (default: false) in order to run multiple filtering steps and only clean up at the end (since that can be quite expensive). However, this may be an example of "YAGNI" - don't implement it until it's actually needed.

Reply:
hmm good point. I wasn't sure what the implications of cleanup can entail e.g., if it can mess up some references if the cleanup is defered for to long etc.
Your call here :)

Reply:
Ok, reading your comment below you want to remove: cleanup_affected and always cleanup if something changed. Right?